src/Avalonia.Controls/ItemsControl.cs

```diff
@@ -219,26 +221,43 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
             return new ItemContainerGenerator(this);
         }
 
+
         /// <summary>
         /// Called when new containers are materialized for the <see cref="ItemsControl"/> by its
         /// <see cref="ItemContainerGenerator"/>.
         /// </summary>
         /// <param name="e">The details of the containers.</param>
         protected virtual void OnContainersMaterialized(ItemContainerEventArgs e)
         {
-            var toAdd = new List<ILogical>();
-
             foreach (var container in e.Containers)
             {
                 // If the item is its own container, then it will be added to the logical tree when
                 // it was added to the Items collection.
                 if (container.ContainerControl != null && container.ContainerControl != container.Item)
                 {
-                    toAdd.Add(container.ContainerControl);
+                    if (ItemContainerGenerator.ContainerType == null)
+                    {
+                        var containerControl = container.ContainerControl as ContentPresenter;
+
+                        if (containerControl != null)
+                        {
+                            ((ISetLogicalParent)containerControl).SetParent(this);
+                            containerControl.SetValue(TemplatedParentProperty, null);
+
+                            containerControl.UpdateChild();
+
+                            if (containerControl.Child is ILogical)
+                            {
+                                LogicalChildren.Add(containerControl.Child as ILogical);
```

Comment:
You've already checked if `containerControl.Child` is an `ILogical` above, so this is unnecessarily casting twice.